loglist/logfilter_test.go

```diff
@@ -0,0 +1,119 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package loglist
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/certificate-transparency-go/testdata"
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509util"
+
+	"github.com/mohae/deepcopy"
+)
+
+func subLogList(t *testing.T, logURLs map[string]bool) LogList {
+	t.Helper()
+	var ll LogList
+	ll.Operators = deepcopy.Copy(sampleLogList.Operators).([]Operator)
+	for _, l := range sampleLogList.Logs {
+		if logURLs[l.URL] {
+			ll.Logs = append(ll.Logs, l)
+		}
+	}
+	return ll
+}
+
+func TestActiveLogs(t *testing.T) {
+	tests := []struct {
+		name string
+		in   LogList
+		want LogList
+	}{
+		{
+			name: "Sample",
+			in:   sampleLogList,
+			want: subLogList(t, map[string]bool{"ct.googleapis.com/aviator/": true, "ct.googleapis.com/icarus/": true, "ct.googleapis.com/rocketeer/": true, "ct.googleapis.com/racketeer/": true}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got := test.in.ActiveLogs()
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("Extracting active logs out of %v. \nExpected %v \nwhile got %v.", test.in, test.want, got)
+			}
+		})
+	}
+}
+
+func certChain(t *testing.T) []*x509.Certificate {
+	t.Helper()
+	cert, _ := x509util.CertificateFromPEM([]byte(testdata.TestPreCertPEM))
+	caCert, _ := x509util.CertificateFromPEM([]byte(testdata.CACertPEM))
+	var chain []*x509.Certificate
+	chain = append(chain, cert)
+	chain = append(chain, caCert)
+	return chain
+}
+
+func singleCert(t *testing.T) []*x509.Certificate {
+	t.Helper()
+	cert, _ := x509util.CertificateFromPEM([]byte(testdata.TestPreCertPEM))
+	var chain []*x509.Certificate
+	chain = append(chain, cert)
+	return chain
+}
+
+func artificialRoots(t *testing.T) LogRoots {
+	m := make(LogRoots)
+	caCert, _ := x509util.CertificateFromPEM([]byte(testdata.CACertPEM))
+	m["log.bob.io"] = []*x509.Certificate{caCert}
+	return m
+}
+
+func TestCompatible(t *testing.T) {
+	tests := []struct {
+		name  string
+		in    LogList
+		chain []*x509.Certificate
+		roots LogRoots
+		want  LogList
+	}{
+		{
+			name:  "RootedChain",
+			in:    sampleLogList,
+			chain: certChain(t),
+			roots: artificialRoots(t),
+			want:  subLogList(t, map[string]bool{"log.bob.io": true}),
+		},
+		{
+			name:  "UnRootedChain",
+			in:    sampleLogList,
+			chain: singleCert(t),
+			roots: artificialRoots(t),
+			want:  subLogList(t, map[string]bool{}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got := test.in.Compatible(test.chain, test.roots)
+			if !reflect.DeepEqual(got, test.want) {
```

Comment:
There are packages you can use to provide a nice diff of `got` and `test.want`, e.g. https://godoc.org/github.com/kr/pretty#Diff

Reply:
For consistency, probably best to stick to https://github.com/kylelemons/godebug/pretty as it's what we use elsewhere.

Reply:
Substituted with pretty-printing.